File: exercises/day18-Risky-Sleigh/kotlin/src/main/kotlin/core/control/ControlSystem.kt
package core.control

import external.deer.Reindeer
import external.stable.MagicStable

class ControlSystem {
    //The Xmas spirit is 40 magic power unit
    private val xmasSpirit = 40
    private val dashboard = Dashboard()
    private val magicStable = MagicStable()
    private var reindeerPowerUnits = bringAllReindeers()
    var status: SleighEngineStatus = SleighEngineStatus.OFF
    var action: SleighAction = SleighAction.PARKED
    private var controlMagicPower = 0f
    private var availableSpecialAmplifiers = mapOf(
        1 to AmplifierType.DIVINE,
        2 to AmplifierType.BLESSED,
        3 to AmplifierType.BLESSED
    )

    init {
        reindeerPowerUnits = bringAllReindeers()
    }

    private fun bringAllReindeers(): List<ReindeerPowerUnit> {
        return BestMagicalPerformancePowerUnitFactory(
            magicStable.allReindeers,
            availableSpecialAmplifiers
        ).bringAllReindeers()
    }

    fun startSystem() {
        dashboard.displayStatus("Starting the sleigh...")

        status = SleighEngineStatus.ON
        dashboard.displayStatus("System ready.")
    }

    @Throws(ReindeersNeedRestException::class, SleighNotStartedException::class)
    fun ascend() {
        if (status == SleighEngineStatus.ON) {
            for (reindeerPowerUnit in reindeerPowerUnits) {
                controlMagicPower += reindeerPowerUnit.harnessMagicPower()
            }

            if (checkReindeerStatus()) {
                dashboard.displayStatus("Ascending...")
                action = SleighAction.FLYING
                controlMagicPower = 0f
            } else throw ReindeersNeedRestException()
        } else {
            throw SleighNotStartedException()
        }
    }

    @Throws(SleighNotStartedException::class)
    fun descend() {
        if (status == SleighEngineStatus.ON) {
            dashboard.displayStatus("Descending...")
            action = SleighAction.HOVERING
        } else throw SleighNotStartedException()
    }

    @Throws(SleighNotStartedException::class)
    fun park() {
        if (status == SleighEngineStatus.ON) {
            dashboard.displayStatus("Parking...")

            for (reindeerPowerUnit in reindeerPowerUnits) {
                // The reindeer rests so the times to harness his magic power resets
                reindeerPowerUnit.reindeer.timesHarnessing = 0
            }

            action = SleighAction.PARKED
        } else throw SleighNotStartedException()
    }

    fun stopSystem() {
        dashboard.displayStatus("Stopping the sleigh...")

        status = SleighEngineStatus.OFF
        dashboard.displayStatus("System shutdown.")
    }

    private fun checkReindeerStatus() = controlMagicPower >= xmasSpirit
}

File: exercises/day18-Risky-Sleigh/ts/src/control.core/controlSystem.ts
import {Dashboard} from "./dashboard";
import {MagicStable} from "../external/stable/magicStable";
import {ReindeerPowerUnit} from "./reindeerPowerUnit";
import {SleighAction, SleighEngineStatus} from "./enums";
import {Reindeer} from "../external/deer/reindeer";
import {ReindeersNeedRestException, SleighNotStartedException} from "./exceptions";
import {AmplifierType} from "./amplifierType";
import {BestMagicalPerformancePowerUnitFactory} from "./bestMagicalPerformancePowerUnitFactory";

export class ControlSystem {
    private XmasSpirit = 40;
    private dashboard: Dashboard;
    private magicStable = new MagicStable();
    private reindeerPowerUnits: ReindeerPowerUnit[];
    public status: SleighEngineStatus;
    public action: SleighAction;
    private controlMagicPower = 0;
    private readonly availableSpecialAmplifiers = new Map<number, AmplifierType>([
        [1, AmplifierType.DIVINE],
        [2, AmplifierType.BLESSED],
        [3, AmplifierType.BLESSED]
    ]);

    constructor() {
        this.dashboard = new Dashboard();
        this.reindeerPowerUnits = this.bringAllReindeers();
    }

    private bringAllReindeers(): ReindeerPowerUnit[] {
        return new BestMagicalPerformancePowerUnitFactory(
            this.magicStable.getAllReindeers(),
            this.availableSpecialAmplifiers
        ).bringAllReindeers();
    }

    public startSystem() {
        this.dashboard.displayStatus("Starting the sleigh...");
        this.status = SleighEngineStatus.ON;
        this.dashboard.displayStatus("System ready.");
    }

    public ascend() {
        if (this.status === SleighEngineStatus.ON) {
            this.reindeerPowerUnits.forEach(unit => {
                this.controlMagicPower += unit.harnessMagicPower();
            });

            if (this.checkReindeerStatus()) {
                this.dashboard.displayStatus("Ascending...");
                this.action = SleighAction.FLYING;
                this.controlMagicPower = 0;
            } else {
                throw new ReindeersNeedRestException();
            }
        } else {
            throw new SleighNotStartedException();
        }
    }

    public descend() {
        if (this.status === SleighEngineStatus.ON) {
            this.dashboard.displayStatus("Descending...");
            this.action = SleighAction.HOVERING;
        } else {
            throw new SleighNotStartedException();
        }
    }

    public park() {
        if (this.status === SleighEngineStatus.ON) {
            this.dashboard.displayStatus("Parking...");

            this.reindeerPowerUnits.forEach(unit => {
                unit.reindeer.timesHarnessing = 0;
            });

            this.action = SleighAction.PARKED;
        } else {
            throw new SleighNotStartedException();
        }
    }

    public stopSystem() {
        this.dashboard.displayStatus("Stopping the sleigh...");
        this.status = SleighEngineStatus.OFF;
        this.dashboard.displayStatus("System shutdown.");
    }

    private checkReindeerStatus(): boolean {
        return this.controlMagicPower >= this.XmasSpirit;
    }
}